XenModel/Actions/Updates/TokenManager.cs

```diff
@@ -30,9 +30,191 @@
+            var clientId = configProvider.FileServiceClientId;
+            var credential = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{username}:{clientId}"));
+
+            var httpWebRequest = (HttpWebRequest)WebRequest.Create(url);
+            httpWebRequest.Headers.Add("Authorization", $"Basic {credential}");
+            httpWebRequest.ContentType = "application/json";
+            httpWebRequest.Method = "POST";
+            httpWebRequest.Proxy = configProvider.GetProxyFromSettings(null, false);
+
+            using (var requestStream = httpWebRequest.GetRequestStream())
+            using (var streamWriter = new StreamWriter(requestStream))
+                streamWriter.Write($"{{\"expiration\": {ExpiryInMinutes}}}");
+
+            try
+            {
+                using (var httpResponse = (HttpWebResponse)httpWebRequest.GetResponse())
+                using (var responseStream = httpResponse.GetResponseStream())
+                {
+                    if (responseStream == null || httpResponse.StatusCode != HttpStatusCode.Created)
+                        throw new WebException();
+
+                    log.Info("Account authenticated. Created token.");
+
+                    using (var streamReader = new StreamReader(responseStream))
+                    {
+                        var json = streamReader.ReadToEnd();
+                        _token = new JavaScriptSerializer().Deserialize(json, typeof(FileServiceToken)) as FileServiceToken;
+                    }
+                }
+            }
+            catch (WebException ex)
+            {
+                if (!(ex.Response is HttpWebResponse response))
+                    throw new WebException(Messages.FILESERVICE_AUTHENTICATE_ERROR);
+
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.BadRequest:
+                        log.Error("Could not authenticate account (400 Bad request). A field provided was invalid");
+                        throw new Exception(string.Format(Messages.STRING_SPACE_STRING,
+                            Messages.FILESERVICE_AUTHENTICATE_ERROR, Messages.FILESERVICE_ERROR_400));
+
+                    case HttpStatusCode.Unauthorized:
+                        log.Error("Could not authenticate account (401 Unauthorized). Client ID may be invalid or revoked.");
+                        throw new Exception(string.Format(Messages.STRING_SPACE_STRING,
+                            Messages.FILESERVICE_AUTHENTICATE_ERROR, Messages.FILESERVICE_ERROR_401));
+
+                    case HttpStatusCode.InternalServerError:
+                        log.Error("Could not authenticate account (500 Internal server error). Could not get session token or configuration object was null.");
+                        throw new Exception(string.Format(Messages.STRING_SPACE_STRING,
+                            Messages.FILESERVICE_AUTHENTICATE_ERROR, Messages.FILESERVICE_ERROR_500));

```

Comment:
In this case, it is more accurate to say that `The authentication server has encountered a situation it doesn't know how to handle`, since 500 can be thrown for all kinds of reasons.

The issue is that it might delay future debugging in case a 500 is being thrown for reasons not related to the session token or the configuration object.